### PR TITLE
Spaceteam proxy stuff

### PR DESCRIPTION
**Pull request by jasonsliu**
According to the TA, we need to submit this tonight.
Thank you!

### CODE REVIEW

File: spaceteam_proxy_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    result_type result = consume(c);
+    if (result == result_type::bad) {
+      return bad_request;
+    }
+  }
+  return OK;
+}
+
+
+SpaceteamProxyHandler::result_type SpaceteamProxyHandler::consume(char input){
```
Comment (nikhilshridhar) [inline spaceteam_proxy_handler.cpp]: Did you guys get this function from an outside boost library? If so you should cite a source for it at the top.

File: spaceteam_proxy_handler.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#define SPACETEAM_PROXY_HANDLER_HPP
+
+#include "request_handler.hpp"
+
+namespace http {
+namespace server {
+class SpaceteamProxyHandler : public RequestHandler {
+public:
+	SpaceteamProxyHandler();
+  virtual ~SpaceteamProxyHandler();
```
Comment (nikhilshridhar) [inline spaceteam_proxy_handler.hpp]: Can you fix the indentation here for the virtual destructor?

File: spaceteam_proxy_handler.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  Headers headers;
+  std::string body;
+
+  // the following are used for parsing the response
+
+  enum result_type { good, bad, indeterminate };
+
+  Status ParseResponse(const std::string& raw_response);
+  result_type consume(char input);
+
```
Comment (nikhilshridhar) [inline spaceteam_proxy_handler.hpp]: Again, if you got this function from an outside boost library, can you please cite it at the top?

File: spaceteam_proxy_handler_test.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  EXPECT_EQ(status_, RequestHandler::Status::OK);
+}
+
+TEST_F(ProxyHandlerTest, NullResponse) {
+  EXPECT_EQ(proxy_handler_.HandleRequest(*request_, nullptr), RequestHandler::Status::ERROR);
+}
+
+TEST_F(ProxyHandlerTest, HandleProperRequestTest) {
+  EXPECT_EQ(proxy_handler_.HandleRequest(*request_, &response_), RequestHandler::Status::OK);
+}
```
Comment (nikhilshridhar) [inline spaceteam_proxy_handler_test.cpp]: Can you make sure for all the new files that you added, you do a single tab of indentation? That is how we would like the format to be. I've seen in some of your files that you have a tab in some of them and a single space in others. Just make sure all the indentations are single tabbed.

File: Makefile [hunk 5]
```diff
@@ -1,7 +1,7 @@
 CXX=g++
 CXXOPTIMIZE= -O2
 CXXFLAGS= -g -Wall -pthread -std=c++11
-SRCFILES = server.cpp response.cpp request.cpp request_parser.cpp config.cc config_parser.cc request_handler.cpp echo_handler.cpp file_handler.cpp not_found_handler.cpp status_handler.cpp server_stats.cpp proxy_handler.cpp
+SRCFILES = server.cpp response.cpp request.cpp request_parser.cpp config.cc config_parser.cc request_handler.cpp echo_handler.cpp file_handler.cpp not_found_handler.cpp status_handler.cpp server_stats.cpp proxy_handler.cpp spaceteam_proxy_handler.cpp
```
Comment (nikhilshridhar) [inline Makefile]: Also, make sure in the Makefile that you add your unit-test file to the test target and your proxy_handler cpp file in the run_coverage target. Look at how the other team who did their proxy handler on ours did this.

**Review by nikhilshridhar — CHANGES_REQUESTED** [review]

File: spaceteam_proxy_handler.cpp (hunk 1 above)
Comment (jasonsliu) [reply spaceteam_proxy_handler.cpp]: Good point, we put the reference in our own repository, but forgot it here.

File: spaceteam_proxy_handler.hpp (hunk 3 above)
Comment (jasonsliu) [reply spaceteam_proxy_handler.hpp]: We'll let the citation live in the .cpp file

File: spaceteam_proxy_handler.cpp [hunk 6]
```diff
@@ -38,324 +40,324 @@ RequestHandler::Status SpaceteamProxyHandler::HandleRequest(const Request& req,
...
-    std::string request = "GET " + relative_uri + " HTTP/1.1\r\n"
-                        + "Host: " + host + ":" + port + "\r\n"
-                        + "Connection: close\r\n\r\n";
-    boost::asio::write(sock, 
-                       boost::asio::buffer(request, request.length()));
+		// send request to host
+		std::string request = "GET " + relative_uri + " HTTP/1.1\r\n"
+												+ "Host: " + host + ":" + port + "\r\n"
+												+ "Connection: close\r\n\r\n";
+		boost::asio::write(sock,
```
Comment (creator23) [inline spaceteam_proxy_handler.cpp]: Why is the spacing here this way?

File: spaceteam_proxy_handler.cpp [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+		result_type result = consume(c);
+		if (result == result_type::bad) {
+			return bad_request;
+		}
+	}
+	return OK;
+}
+
+
+SpaceteamProxyHandler::result_type SpaceteamProxyHandler::consume(char input){
```
Comment (creator23) [inline spaceteam_proxy_handler.cpp:130]: This code is duplicated, please check request_parser.cpp

**Review by creator23 — CHANGES_REQUESTED** [review]
Hi, some of the code, specifically the consume code that you wrote in the proxy_handler file is already present in our codebase, please see request_parser.cpp and modify your code to use this since we would want to avoid redundant duplication.

File: spaceteam_proxy_handler.cpp (hunk 7 above)
Comment (jasonsliu) [reply spaceteam_proxy_handler.cpp:130]: It's not duplicated! Our's is different from yours, though a decent amount is the same.

**Comment by jasonsliu** [comment]
Hi Team07, it's getting pretty close to our deadline. Are there any more issues we need to address?

**Review by creator23 — APPROVED** [review]

**Review by creator23 — APPROVED** [review]
Code looks good, we'll merge it in soon.

**Comment by jasonsliu** [comment]
It's past the deadline, could we please merge it...

**Pull request merged**